CoEpi/native/LibResult.swift

```diff
@@ -0,0 +1,50 @@
+import Foundation
+
+struct LibResult<T: Decodable>: Decodable {
+    let status: Int
+    let data: T?
+    let error_message: String?
+
+    func isSuccess() -> Bool {
+        (200...299).contains(status)
+    }
+}
+
+// TODO better way to decode LibResult when it doesn't have data. Maybe don't use Decodable.
+typealias ArbitraryType = String
+
+extension LibResult {
+
+    func toResult() -> Result<T, ApiError> {
+        if isSuccess() {
+            if let data = data {
+                return .success(data)
+            } else {
+                return .failure(.error(message: "Unexpected: Library result success but no data: \(self)"))
+            }
+        } else {
+            return .failure(.error(message: "Lib error result: \(self)"))
+        }
+    }
+
+    func toVoidResult() -> Result<(), ApiError> {
+        if isSuccess() {
+            return .success(())
+        } else {
+            return .failure(.error(message: "Lib error result: \(self)"))
+        }
+    }
+}
+
+extension Unmanaged where Instance == CFString {
+
+    func toLibResult<T>() -> LibResult<T> {
+        let resultValue: CFString = takeRetainedValue()
+        let resultString = resultValue as String
+
+        // TODO review safety of utf-8 force unwrap
+        let data = resultString.data(using: .utf8)!
+        let decoder = JSONDecoder()
+        return try! decoder.decode(LibResult<T>.self, from: data)
```

Comment:
Should we handle errors in decoding?

Reply:
yup

Reply:
I decided to return a `LibResult` with 502 (see note in code) instead of `Result<LibResult...>` to simplify the caller code.